### PR TITLE
Badger2 integration

### PR DESCRIPTION
**Pull request by gammazero**
This is a provisional PR and should not be merged until dependencies are merged into their master branches.

- Add badger2 plugin
- Update docs to describe badger2
- Add tests for badger2
- Set dependencies to pending PRs with necessary updates

### CODE REVIEW

File: docs/datastores.md
Comment (aschmahmann) [inline docs/datastores.md:66]: Should add a compression flag here (and in the various config locations) if it's not an init-only parameter.

File: test/sharness/t0025-datastores.sh
Comment (aschmahmann) [inline test/sharness/t0025-datastores.sh:21]: BITS can be removed here and above, as they're no longer needed.

File: plugin/plugins/badger2ds/badger2ds.go
Comment (aschmahmann) [inline plugin/plugins/badger2ds/badger2ds.go:2]: If Badger isn't going to fix their x86 issues (especially the compilation error), then we should split this package into two. One with a build directive for `386` (or easier just calling the file badger2ds_386.go) that just returns errors and the other with a `// +build !386` directive.

That should mostly get things to work, although to get tests to pass we'll also need to add an exception to not run the badger2 sharness tests on 386.

**Review by aschmahmann — CHANGES_REQUESTED** [review]

File: plugin/plugins/badger2ds/badger2ds.go
Comment (gammazero) [reply plugin/plugins/badger2ds/badger2ds.go:2]: Badger has fixed the build failures, which allows out builds to work.  Config changes on our side allow our tests to pass. So, we should be ok, and we are completely avoiding the mmap problem now so we do not need to issue the warning about 32bit.  Badger should fix their 32bit tests, but we are avoiding the cause of those failures (mmap + 32bit).

File: docs/datastores.md
Comment (gammazero) [reply docs/datastores.md:66]: Maybe this should be a separate PR.  If we do this then we should also add in options for encryptions and cache size.  The cache size is needed, because the badger2 documentation recommends using a cache if either compression or encryption is enabled. And if those are enabled, then it may be responsible to add some IPFS tests with those options.

File: test/sharness/t0025-datastores.sh
Comment (gammazero) [reply test/sharness/t0025-datastores.sh]: Done

File: docs/datastores.md
Comment (gammazero) [reply docs/datastores.md:68]: Done

**Comment by aschmahmann** [comment]
@gammazero your go.mod files are conflicting with master. Could you rebase when you get a chance?

File: repo/fsrepo/config_test.go
Comment (aschmahmann) [inline repo/fsrepo/config_test.go:323]: would using reflect.Typeof(ds).PkgPath() help disambiguate here?

File: docs/datastores.md
Comment (aschmahmann) [inline docs/datastores.md:70]: WDYT about having some value for each of these signifying the default (e.g. in https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#types we have some default values). This would allow us (the implementers) to change the default compression and blockCacheSize without users needing to be change their configs. Of course, if they wanted to fix these parameters they could explicitly put them in the config file.

My thinking is to use `null or missing` to mean "apply the default value" as we do with Flag and Priority and then define what our current default values are (no compression, no blockCacheSize).

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Looking good. Added a couple small comments and a suggestion on the compression + blockCacheSize parameters.

File: docs/datastores.md
Comment (gammazero) [reply docs/datastores.md:70]: I would prefer that anytime a string parameter is missing or has a null (or empty `''"`) value, that this means use the default of our choosing (which could change with updates) -- _when appropriate_.  There are some cases where a value cannot change with an upgrade, such as a filesystem location. So, for these we cannot use a default for a missing or null. In those cases, we must supply a default value at initialization time, and report and error if the value is ever unspecified.

So, I do not think we can make a general statement that all missing/null string values get set to some default. This needs to be done on a case-by-case basis.  Some strings can be missing/null and a reasonable default will be used, and for others an initial value is provided but a value must always be present.

So, to address this, I have amended the description specifically for the new badger2 config values, that omitting the values from the configuration results in choosing the default setting.

File: repo/fsrepo/config_test.go
Comment (gammazero) [reply repo/fsrepo/config_test.go:323]: Unfortunately `PkgPath()` always returns `""` here.  I cannot yet find a clear explanation why, but it appears to do with the way interfaces are embedded.  However, if we phase-out badger1 then this will no longer be an issue.

File: plugin/plugins/badger2ds/badger2ds.go
Comment (gammazero) [reply plugin/plugins/badger2ds/badger2ds.go]: Comment has been clarified to state that missing value results in go-ds-badger2 defaults.  This is to tell developers where to look to see what values will be used.  Mostly, this means using default values from badger/v2, but the point is to give tell developers where the next link in this chain is.

File: docs/datastores.md
Comment (aschmahmann) [inline docs/datastores.md:64]: It looks like both badger and badger2 documentation do not specify that you can set `vlogFileSize`.  @Stebalien this is just an oversight right?

Comment (Stebalien) [reply docs/datastores.md:64]: Yes.

Comment (aschmahmann) [reply docs/datastores.md:64]: @gammazero could you add in the `vlogFileSize` documentation as well?

Comment (gammazero) [reply docs/datastores.md:67]: Done.

**Review by Stebalien — APPROVED** [review]
:rocket:

**Review by aschmahmann — APPROVED** [review]
This looks good to go pending a release of go-ds-badger2 and upgrading go-ipfs-config to v0.10.0

**Comment by aschmahmann** [comment]
This is blocked by the badger folks figuring out if there latest breaking changes will be handled within badger v2 or whether there will be a badger v3. The go-ipfs-config PR adding badger 2 support has been reverted until this is resolved.

**Comment by dokterbob** [comment]
@aschmahmann There's now a Badger v3 https://github.com/dgraph-io/badger/releases/tag/v3.2011.1

However, go-ds-badger2 is actively maintained. Requirements mentioned by you have been met a long time ago and, even as an experimental feature, it could gain great benefit and insight to have something *less broken* to test with. This might also yield useful feedback while a move towards badger 3 is made.

We currently have a badgerdb in IPFS which is broken in [obvious ways](https://github.com/ipfs/go-ds-badger/issues/54#issuecomment-819866098) making it unsuitable for production.

**Comment by dokterbob** [comment]
In short; all that's necessary to have a badger v2, based on actively maintained code, in master, is to resolve these merge conflicts. Getting v3 support is probably a fair amount of work and might delay having something working by several months/up to a year. 

It would be great to have this in user's hands!

**Comment by godcong** [comment]
I tried to use V3, but it will panic in some cases.

**Comment by dokterbob** [comment]
@aschmahmann Poke!

**Comment by dokterbob** [comment]
At ipfs-search.com we currently have a [script](https://github.com/ipfs-search/ipfs-search-deployment/blob/main/roles/ipfs-gc/templates/ipfs_gc.sh) running from a cron job which stops IPFS, flattens the datastore using the badger binary and then restarts IPFS.

I know you've got a lot on your hands, but it would be great to see some progress on this one!

**Comment by dokterbob** [comment]
Poke!

**Comment by feld** [comment]
flatfs is not good when you've got 10s to 100s of gigs in IPFS. One of my use cases can survive with a manual flatten as suggested by @dokterbob but it would be nice if we had badger v2 or v3 to work with

**Comment by feld** [comment]
@dokterbob I have a branch with Badger2 working on IPFS v0.14.0 and garbage collection does work properly

https://github.com/feld/kubo/tree/feld/badger2-rebase-v0.14.0

**Comment by lidel** [comment]
Closing this one – iirc long term we will more likely introduce pebble (https://github.com/ipfs/kubo/issues/10347) and short-to-mid term discourage use of badgerv1 (https://github.com/ipfs/kubo/pull/10517)

**Pull request closed without merging**
